File: arduinobot_ws/src/arduinobot_firmware/arduinobot_firmware/simple_serial_receiver.py
#!/usr/bin/env python3
import serial
import time
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class SimpleSerialReceiver(Node):
    def __init__(self):
        super().__init__("simple_serial_receiver")

        self.declare_parameter("port", "/dev/ttyUSB0")
        self.declare_parameter("baudrate", 115200)

        self.port_ = self.get_parameter("port").value
        self.baudrate_ = self.get_parameter("baudrate").value

        self.pub_ = self.create_publisher(String, "simple_receiver", 10)
        self.arduino_ = serial.Serial(port=self.port_, baudrate=self.baudrate_, timeout=0.1)

    def execute(self):
        while rclpy.ok() and self.arduino_.is_open:
            data = self.arduino_.readline()

            try:
                data.decode("utf-8")
            except:
                continue
            
            self.get_logger().info("New data available on serial, publishing on ROS 2 topic")
            msg = String()
            msg.data = str(data)
            self.pub_.publish(msg)
            time.sleep(0.01)


def main():
    rclpy.init()

    simple_serial_transmitter = SimpleSerialReceiver()
    simple_serial_transmitter.execute()
    rclpy.spin(simple_serial_transmitter)
    
    simple_serial_transmitter.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: arduinobot_ws/src/arduinobot_firmware/src/simple_serial_receiver.cpp
#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

#include <chrono>
#include <thread>


class SimpleSerialReceiver : public rclcpp::Node
{
public:
  SimpleSerialReceiver() : Node("simple_serial_receiver")
  {
    declare_parameter<std::string>("port", "/dev/ttyUSB0");
    declare_parameter<int>("baudrate", 115200);

    port_ = get_parameter("port").as_string();
    baudrate_ = get_parameter("baudrate").as_int();

    pub_ = create_publisher<std_msgs::msg::String>("simple_receiver", 10);
  }

  void execute()
  {
    while(rclcpp::ok())
    {
        auto message = std_msgs::msg::String();
        message.data = "data";
        pub_->publish(message);
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
  }

private:
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
  std::string port_;
  int baudrate_;
};


int main(int argc, char* argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SimpleSerialReceiver>();
  node->execute();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
